Refactor

movermeyer opened the pull request:
* Fixed the consistency issues (Solves #3 and Solves #10).
* Removed `format_string` (Solves #2)
* Refactored the internals to have greater separation of concern.
* The reason(s) for the reports is now passed to format_callback. Allows for advanced print customization. (Solves #9).
* every_n_seconds_idle now reports the idle time correctly, without the processing time included.

movermeyer committed:
Made every_x_seconds consistent with every_n_records.

Fixes #10

movermeyer committed:
Made `last_iteration` and `ignore_first_iteration` consistently named.

Fixes #3

movermeyer committed:
Removed `format_string` param. Use lambda with `format_callback` param instead.

movermeyer committed:
- Refactored the internals to have greater separation of concern.
- The reason(s) for the reports is now passed to format_callback. Allows for advanced print customization. Solves #9.
WARNING: This changed the signature of format_callback.
- every_n_seconds_idle now reports the idle time correctly, without the processing time included.

movermeyer committed:
Added section for Mypy configuration

movermeyer committed:
Merge pull request #1 from movermeyer/refactor_reports

Refactor reports

movermeyer committed:
Merge pull request #2 from movermeyer/remove_format_string

Remove format string

movermeyer committed:
Fixed tests to use the new format_callback syntax.

NickG123 commented on progress_tracker/progress_tracker.py:
```diff
@@ -69,77 +71,85 @@ def __init__(self, iterable: Iterable[T],
+            for item in self.iterable:
+                # Has it been a while since we last saw a record?
+                if self.idle_timeout is not None and self.idle_timeout.is_overdue():
+                    self.raise_report(set(["every_n_seconds_idle"]))
```
This could raise multiple reports for a single item - is that desired?

movermeyer replied:
Yes. In the case where there has been an idle period, I want a report of the amount of idle time.
I recognize that this could result in:

"Progress: 3/100"
"Idle for 3 minutes. Progress: 3/100"
"Progress: 4/100"

Perhaps I should clean up those assertions in the README, but yes this is expected behaviour.

movermeyer replied:
I mean, I could change it so that it does this instead:

"Progress: 3/100"
"Progress: 4/100 (after being idle for 3 minutes)"

This would maintain the "At most one report per record" mantra. I guess it becomes a question of "How valuable is that invariant?"

NickG123 replied:
I think that's more appealing, but as long as it's well documented it doesn't really matter

movermeyer replied:
Fixed. Made the invariants more explicit in README, and made the code observe them.

NickG123 commented on README.rst:
```diff
@@ -56,8 +56,8 @@ By changing the parameters passed to ``track_progress``, you can customize how f
         every_n_seconds_idle: Optional[float] = None, # Report every n seconds, but only if there hasn’t been any progress. Useful for infinite streams
-        ignore_first_iteration: bool = True, # Don’t report on the first iteration
-        last_iteration: bool = False # Report after the last iteration
+        report_first_record: bool = False, # Report after the first record
```
Changing this name actually changes meaning.  I think you need to make a code change in order for `track_progress(x, report_first_record=True)` to actually print something for the first record

movermeyer replied:
You're right. I've changed the code to fix it.

It never actually worked as described. For example, it would still report on first iteration if any of the timers triggered, which is not what the name, nor the code comment, suggests.

Also, `last_iteration` never reported after the `StopIteration` on unbounded Iterables.

I've fixed both cases. Good catch.

movermeyer replied:
Also, if there is only a single record, and both `report_first_record=True` and `report_last_record=True`, then this is another case where a single record causes multiple reports.

(On unbounded iterables, because of the way I implemented it. This was a mistake. I will fix it.

movermeyer replied:
Fixed. Made the invariants more explicit in README, and made the code observe them.

NickG123 commented on progress_tracker/progress_tracker.py:
```diff
@@ -1,21 +1,23 @@
-    total = kwargs["total"]
-    if total is None or i == total:
-        format_string = "{i} in {time_taken}"
+def default_format_callback(report: Dict[str, Any], reasons: Set[str]) -> str:
```
As a side note, I don't think the contents of `report` are documented anywhere

movermeyer replied:
Yeah...They should be. Blech.

Alright. I'll do it.

movermeyer replied:
Done.

NickG123 commented on progress_tracker/progress_tracker.py:
```diff
@@ -69,77 +71,85 @@ def __init__(self, iterable: Iterable[T],
+        reasons_to_report: Set[str] = set()
         if self.timeout is not None and self.timeout.is_overdue():
-            should_report = True
+            reasons_to_report.add("every_n_seconds")
```
I think it makes sense for `reasons_to_report` to be a list of enum instances, not strings.  Then when writing a custom handler a user can import the enum and use it instead of magic strings in their handler.

movermeyer replied:
Fair. Will fix.

movermeyer replied:
One reason to keep it as strings is to allow for easy subclassing of ProgressTracker.

If you subclassed `ProgressTracker` to add a new condition, you would override `should_report`, but you [cannot extend the Enum](https://docs.python.org/3/library/enum.html#restricted-subclassing-of-enumerations) to add your new reason.

NickG123 replied:
Then perhaps constants.  It just feels cleaner to be able to import something from this library

movermeyer committed:
Fixed cases where report_first_record and report_last_record weren't reporting correctly.

movermeyer committed:
Made report creation invariants explicit.
Moved check for `every_n_seconds_idle` back into `should_report`, to avoid multiple reports for a single record.
Unified `report_last_record` for all iterables. Fixed issue where `report_last_record` would fire on empty unbounded iterables.

movermeyer committed:
Fixed Mypy complaint

movermeyer committed:
Made the reasons constants. This allows for easy user extension and protection from silent failure if available reasons were to change.

movermeyer committed:
Documented available report values.

NickG123 commented:
Looks good
